File: benchmarks/pandas/bench_filter_series.py
"""Benchmark: Series.filter — filter Series index labels by items/like/regex"""
import json
import time
import pandas as pd

N = 100_000
WARMUP = 3
ITERATIONS = 10

labels = [f"label_{i}" for i in range(N)]
values = [i * 0.5 for i in range(N)]
s = pd.Series(values, index=labels)

keep_items = [f"label_{i * 100}" for i in range(1_000)]

for _ in range(WARMUP):
    s.filter(items=keep_items)
    s.filter(like="label_5")

start = time.perf_counter()
for _ in range(ITERATIONS):
    s.filter(items=keep_items)
    s.filter(like="label_5")
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "filter_series",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))

File: benchmarks/tsb/bench_filter_series.ts
/**
 * Benchmark: filterSeries — filter Series index labels by items/like/regex
 * Outputs JSON: {"function": "filter_series", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, filterSeries } from "../../src/index.ts";

const N = 100_000;
const WARMUP = 3;
const ITERATIONS = 10;

// Series with string labels: "label_0", "label_1", ..., "label_N-1"
const labels = Array.from({ length: N }, (_, i) => `label_${i}`);
const values = Array.from({ length: N }, (_, i) => i * 0.5);
const s = new Series<number>({ data: values, index: labels });

// Pre-build a set of 1000 items to keep
const keepItems = Array.from({ length: 1_000 }, (_, i) => `label_${i * 100}`);

for (let i = 0; i < WARMUP; i++) {
  filterSeries(s, { items: keepItems });
  filterSeries(s, { like: "label_5" });
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  filterSeries(s, { items: keepItems });
  filterSeries(s, { like: "label_5" });
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "filter_series",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);
